File: number_of_islands.cc
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        if (grid.size() == 0) {
            return 0; 
        }
        m = grid.size();
        if (grid[0].size() == 0) {
            return 0; 
        }
        n = grid[0].size();
        int count = 0; 
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                 if (grid[i][j] != '1') {
                      continue; 
                 } else {
                      count += 1;
                      dfs(grid, i, j);
                 }
            } 
        }
        return count;
    }
    void dfs(vector<vector<char>>& grid, int i, int j) {
        if (i < 0 || i >= m || j < 0 || j >= n) {
            return; 
        } 
        if (grid[i][j] == '1') {
             grid[i][j] = '0';
             dfs(grid, i - 1, j);
             dfs(grid, i, j - 1);
             dfs(grid, i + 1, j);
             dfs(grid, i, j + 1);
        }
    }
private:
    int m;
    int n;
};

File: python/number_of_islands.py
class Solution:
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        m = len(grid)
        if m == 0:
            return 0
        n = len(grid[0])
        result = 0
        dirs = [[-1, 0], [1, 0], [0, -1], [0, 1]]
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    result += 1
                    self.dfs(grid, i, j, dirs)
        return result


    def dfs(self, grid, i, j, dirs):
        if i >= 0 and i < len(grid) and j >= 0 and j < len(grid[0]) and \
           grid[i][j] == '1':
            grid[i][j] = '#'
            for d in dirs:
                x = i + d[0]
                y = j + d[1]
                self.dfs(grid, x, y, dirs)
